### ISSUE TITLE
Content-Type returned by JSON is application/text when used in special middlewares

### ISSUE THREAD

**Issue by aharbis** [uptrace/bunrouter#78]
## 💡 The feature or bug you are proposing

Using `bunrouter.JSON` in handlers passed to either `WithNotFoundHandler` or `WithMethodNotAllowedHandler` results in `application/text` Content-Type header.

## 📄 The description of the bug or the logic behind your proposal

Constructing my router like this:

```go
router := bunrouter.New(
	bunrouter.WithMiddleware(reqLogMiddleware),
	bunrouter.WithMiddleware(errorHandlerMiddleware),
	bunrouter.WithNotFoundHandler(notFoundHandler),
	bunrouter.WithMethodNotAllowedHandler(methodNotAllowedHandler),
)
```

Both of my `notFoundHandler` and `methodNotAllowedHandler` middlewares send a valid JSON response with:

```go
bunrouter.JSON(w, err)
```

Where `err` is a custom `HTTPError` type with `json` marshaling tags.

The JSON marshaling does work and the response received on the client is something like:

```
{"code":"not_found","message":"no route found for /"}
```

But the Content-Type header received is `application/text`.

## 🚀 The expected result

`bunrouter.JSON` should always set `application/json` as the Content-Type, even when used in special middlewares.

**Comment by aharbis** [comment]
Actually, this may not be scoped to just those middlewares. I think it may be related to the `POST` method instead.

**Comment by aharbis** [comment]
I think I found the root cause, note this comment in [`http.ResponseWriter`](https://pkg.go.dev/net/http#ResponseWriter):

```go
// Changing the header map after a call to WriteHeader (or
// Write) has no effect unless the HTTP status code was of the
// 1xx class or the modified headers are trailers.
```

However in the [bunrouter documentation](https://bunrouter.uptrace.dev/guide/golang-http-error-handling.html), and indeed my code, the following pattern is used:

```go
w.WriteHeader(httpErr.statusCode)
return bunrouter.JSON(w, httpErr)
```

This places the call to `w.Header().Set()` after `w.WriteHeader()`, which prevents the `Content-Type` from being set as expected.

**Comment by vmihailenco** [comment]
Yes, you must be careful with `WriteHeader`. It is not specific to BunRouter and it does not offer a universal solution.

https://bunrouter.uptrace.dev/guide/golang-http-error-handling.html may give you some ideas. Also see [uptrace](https://github.com/uptrace/uptrace/blob/master/pkg/bunapp/app.go#L235-L254) for a more realistic example.
